v4/newrelic/transaction.go

```diff
@@ -0,0 +1,348 @@
+//   ErrorClass() string
+//
+//   // ErrorAttributes sets the errors attributes
+//   ErrorAttributes() map[string]interface{}
+//
+// The newrelic.Error type, which implements these methods, is the recommended
+// way to directly control the recorded error's message, class, stacktrace,
+// and attributes.
+func (txn *Transaction) NoticeError(err error) {}
+
+// AddAttribute adds a key value pair to the transaction event, errors,
+// and traces.
+//
+// The key must contain fewer than than 255 bytes.  The value must be a
+// number, string, or boolean.
+//
+// For more information, see:
+// https://docs.newrelic.com/docs/agents/manage-apm-agents/agent-metrics/collect-custom-attributes
+func (txn *Transaction) AddAttribute(key string, value interface{}) {}
+
+// SetWebRequestHTTP marks the transaction as a web transaction.  If
+// the request is non-nil, SetWebRequestHTTP will additionally collect
+// details on request attributes, url, and method.  If headers are
+// present, the agent will look for distributed tracing headers using
+// Transaction.AcceptDistributedTraceHeaders.
+func (txn *Transaction) SetWebRequestHTTP(r *http.Request) {}
+
+// SetWebRequest marks the transaction as a web transaction.  SetWebRequest
+// additionally collects details on request attributes, url, and method if
+// these fields are set.  If headers are present, the agent will look for
+// distributed tracing headers using Transaction.AcceptDistributedTraceHeaders.
+// Use Transaction.SetWebRequestHTTP if you have a *http.Request.
+func (txn *Transaction) SetWebRequest(r WebRequest) {}
+
+// SetWebResponse allows the Transaction to instrument response code and
+// response headers.  Use the return value of this method in place of the input
+// parameter http.ResponseWriter in your instrumentation.
+//
+// The returned http.ResponseWriter is safe to use even if the Transaction
+// receiver is nil or has already been ended.
+//
+// The returned http.ResponseWriter implements the combination of
+// http.CloseNotifier, http.Flusher, http.Hijacker, and io.ReaderFrom
+// implemented by the input http.ResponseWriter.
+//
+// This method is used by WrapHandle, WrapHandleFunc, and most integration
+// package middlewares.  Therefore, you probably want to use this only if you
+// are writing your own instrumentation middleware.
+func (txn *Transaction) SetWebResponse(w http.ResponseWriter) http.ResponseWriter {
+	return w
+}
+
+// StartSegmentNow starts timing a segment.  The SegmentStartTime returned can
+// be used as the StartTime field in Segment, DatastoreSegment, or
+// ExternalSegment.  The returned SegmentStartTime is safe to use even  when the
+// Transaction receiver is nil.  In this case, the segment will have no effect.
+func (txn *Transaction) StartSegmentNow() SegmentStartTime {
+	if txn == nil {
+		return SegmentStartTime{}
+	}
+	if txn.isEnded() {
+		return SegmentStartTime{}
+	}
+	parent := txn.thread.getCurrentSpan()
```

Comment:
Do you think we could obtain the current span from OTel, without managing it in the shim layer? I think otherwise we might run into problems mixing the Go Agent and OTel instrumentation. I'm thinking about a test case like this:
```go
app, tracer := newTestApp(t)
txn := app.StartTransaction("transaction")

seg1 := txn.StartSegment("seg1")

_, span := tracer.Start(context.Background(), "seg2")
span.End()

seg1.End()
txn.End()
```
In the ideal case, this should link parent `seg2` to `seg1`. But I think managing the current span in the transaction state interferes with that.

Reply:
I think we also should add something similar to the code snippet above as a unit test case.

Reply:
Fantastic point. I will consider this.

Reply:
I had a closer look at this, and I'm actually not sure how far we can get without managing parenting in the shim layer.

When looking at OpenTelemetry Go, I see that the parenting is handling by using context. The parent context is passed to `tracer.Start`, which in turn returns the span and an associated context. This can then be passed on to further `tracer.Start` calls to create children of the context passed in. The currently active span is contained in the context, and is used by `tracer.Start` to set the parent span.

Maybe we have to stay with the solution you came up with. We could then introduce a new `GetContext` method on the transaction (maybe there's a better name), which returns the context related to the currently active segment. The sample above would then look like this:
```go
app, tracer := newTestApp(t)
txn := app.StartTransaction("transaction")

seg1 := txn.StartSegment("seg1")

_, span := tracer.Start(txn.GetContext(), "seg2")
span.End()

seg1.End()
txn.End()
```

Reply:
Yes, this is what I was thinking as well. However instead of a `GetContext` we could just use the already existing `NewContext` method.

```go
app, tracer := newTestApp(t)
txn := app.StartTransaction("transaction")

seg1 := txn.StartSegment("seg1")

ctx := newrelic.NewContext(context.Background(), txn)
_, span := tracer.Start(ctx, "seg2")
span.End()

seg1.End()
txn.End()
```

Reply:
Hmmm, though you might be right about needing a new `GetContext` method. Either way, I think that decision can wait until we begin work on that aspect of the project.

Reply:
Hmm, and we should consider the case where the OTel span is the parent of a newrelic segment. I'm not sure how to handle that. Any ideas?